### PR TITLE
[REVIEW] Memory Manager

### PR DESCRIPTION
**Pull request by harrism**
Initial implementation of Memory Manager for libgdf, called RMM.  RMM by default just wraps cudaMalloc/cudaFree, but if you undefined RMM_USE_CUDAMALLOC it will use an internal memory pool implementation. This currently is based on CNMem, which is now a libgdf submodule.

RMM provides a basic alloc/free/realloc API, along with logging. It also has Python bindings, which provide an API compatible with Numba's `cuda.device_array` so that `device_array` can be allocated through the memory pool rather than direct to CUDAMalloc. 

libgdf also now includes `thrust_rmm_allocator.h`, which defines an allocator that can be used with Thrust `device_vector` and thrust algorithm temporary allocations by passing it to the `thrust::cuda::par(allocator)` execution policy.

This is a work in progress.

### CODE REVIEW

**Comment by scopatz** [comment]
Hi @harrism - Just out of curiosity, what is the reason for creating this?  It seems to touch a lot of the code base.

**Comment by kkraus14** [comment]
> Hi @harrism - Just out of curiosity, what is the reason for creating this? It seems to touch a lot of the code base.

@nsakharnykh I think you can explain this better than I can 😄

**Comment by harrism** [comment]
It's a memory pool allocator. Currently the biggest bottleneck in libgdf is cudaMalloc / cudaFree. A pool allocator is intended to resolve that bottleneck, as well as give us a central location for tracking all memory usage.

**Comment by jrhemstad** [comment]
> Hi @harrism - Just out of curiosity, what is the reason for creating this? It seems to touch a lot of the code base.
@scopatz In short, cudaMalloc/cudaFree are very slow and they are our primary bottleneck right now.

@harrism has created a custom memory manager for us that replaces every memory allocation with a custom memory allocator (which is why it touches so much code). It's based on CNMEM, which basically allocates a huge pool of memory up front with cudaMalloc. All allocations called with `rmm_malloc` are simply given a chunk of memory in this memory pool, and the custom memory manager keeps track of which allocations are using which blocks in the memory pool. When a chunk of memory is freed, it is simple made available to future allocations in the pool.

**Comment by harrism** [comment]
rerun tests

**Comment by pearu** [comment]
Perhaps offtopic: Apache Arrow uses the memory pool concept as well but currently for host memory only: https://arrow.apache.org/docs/cpp/memory__pool_8h_source.html
It might be useful to add the device memory pool support to CUDA enabled Arrow as well.

File: src/memory/memory.cpp [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+#else
+    RMM_CHECK_CNMEM( cnmemFree(ptr, stream) );
+#endif
+	return RMM_SUCCESS;
+}
+
+// Get the offset of ptr from its base allocation
+rmmError_t rmmGetAllocationOffset(ptrdiff_t *offset, void *ptr, cudaStream_t stream)
+{
+#ifdef RMM_USE_CUDAMALLOC
```
Comment (nsakharnykh) [inline src/memory/memory.cpp]: This doesn't seem correct, should it be `#ifndef`?

Comment (jrhemstad) [reply src/memory/memory.cpp]: Yeah, I think @nsakharnykh is right.

Comment (harrism) [reply src/memory/memory.cpp]: This actually popped into my head right after I pushed but I was boarding a flight. That's what I get for programming before boarding.

**Comment by harrism** [comment]
@pearu thanks for the tip -- I will look into that after next week.

**Comment by scopatz** [comment]
I see, thanks for clearing that up. I think the big question for us is how is this envisioned to interact with Arrow, which is what @pearu was hinting at.  Is RMM something that should live outside of libgdf, so that arrow can use it, and then libgdf just uses arrow?

File: python/librmm_cffi/wrapper.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
+import numpy as np
+from numba import cuda 
+import ctypes
+
+class RMMError(Exception):
```
Comment (scopatz) [inline python/librmm_cffi/wrapper.py]: Wouldn't this be better as a subclass of `MemoryError`?

File: python/librmm_cffi/wrapper.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    def __init__(self, errcode, msg):
+        self.errcode = errcode
+        super(RMMError, self).__init__(msg)
+
+
+class _librmm_wrapper(object):
```
Comment (scopatz) [inline python/librmm_cffi/wrapper.py]: Class names should be `CapCase`

File: src/memory/memory_manager.h [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    if (CNMEM_STATUS_CUDA_ERROR == error) \
+        return RMM_ERROR_CUDA_ERROR; \
+    else if (CNMEM_STATUS_INVALID_ARGUMENT == error) \
+        return RMM_ERROR_INVALID_ARGUMENT; \
+    else if (CNMEM_STATUS_NOT_INITIALIZED == error) \
+        return RMM_ERROR_NOT_INITIALIZED; \
+    else if (CNMEM_STATUS_OUT_OF_MEMORY == error) \
+        return RMM_ERROR_OUT_OF_MEMORY; \
+    else if (CNMEM_STATUS_UNKNOWN_ERROR == error) \
+        return RMM_ERROR_UNKNOWN; \
```
Comment (scopatz) [inline src/memory/memory_manager.h]: This seems like it should be a switch-case

File: src/thrust_rmm_allocator.h [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+      if (error != RMM_SUCCESS)
+      {
+        throw thrust::system_error(error, thrust::cuda_category(), "rmm_temp_allocator::deallocate(): rmmFree");
+      }
+    }
+
+  private:
+    cudaStream_t stream;
+};*/
```
Comment (scopatz) [inline src/thrust_rmm_allocator.h]: Please remove dead code

**Comment by harrism** [comment]
The latest failures are due to TravisCI not having libcuda.so. I believe @mike-wendt has a planned fix for this.

File: python/librmm_cffi/wrapper.py (hunk 3 above)
Comment (harrism) [reply python/librmm_cffi/wrapper.py]: This naming is copied from _libgdf_wrapper, which I did not write.

File: python/librmm_cffi/wrapper.py (hunk 2 above)
Comment (harrism) [reply python/librmm_cffi/wrapper.py]: Is GDFError a subclass of anything? I copied this from GDFError.

**Comment by mike-wendt** [comment]
@harrism @nsakharnykh @jrhemstad fixed Travis CI so libcuda is installed and works for building. Both Travis CI and gpuCI report ✅

**Comment by mike-wendt** [comment]
I would recommend merging #171 prior to this commit so we can deal with any issues in the PR instead of master

**Comment by nsakharnykh** [comment]
I believe this is good now, but I don't know why CI failed? @mike-wendt

**Comment by mike-wendt** [comment]
Looking at it now. Probably need to reorder the apt installs. Will test a fix for that

**Comment by mike-wendt** [comment]
LGTM

**Pull request merged**